libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteSmoother.kt

```diff
@@ -0,0 +1,193 @@
+package com.mapbox.navigation.core.replay.route2
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.turf.TurfConstants
+import com.mapbox.turf.TurfMeasurement
+import com.mapbox.turf.TurfMeasurement.EARTH_RADIUS
+import kotlin.math.abs
+import kotlin.math.cos
+import kotlin.math.sin
+import kotlin.math.sqrt
+
+/**
+ * In order to estimate the speed needed to traverse a [DirectionsRoute] we need
+ * a route with locations that are significant for speed.
+ *
+ * This class reduces a list of location coordinates into a list of locations
+ * with direction changes beyond a threshold.
+ */
+internal class ReplayRouteSmoother {
+
+    /**
+     * Take a list of coordinates and return a list of locations where there are significant
+     * changes in bearing. Each location will have a bearing and a distance.
+     */
+    fun smoothRoute(distinctPoints: List<Point>, thresholdMeters: Double): List<ReplayRouteLocation> {
```

Comment:
What about extracting these different blocks of code (separated with new lines) into `private` functions and give them a proper name? This way `smoothRoute` will be easier to read and understand. We should keep `public` functions as clean as possible so they read like pseudocode (i.e. no conditionals, `for`s, etc.).

Reply:
one note, is the entire class is `internal`. so these functions are really meant to be mapbox protected. but yeah going to dig into making this cleaner some more.

Reply:
Yeap yeap I know but let's try to do it for our future us 😉 (also know that when it comes to math is even harder 😬)

Reply:
Broke it into a couple smaller private functions